// tslint/lib/rules/noReferenceRule.js
/* */ 
"use strict";
Object.defineProperty(exports, "__esModule", {value: true});
var tslib_1 = require('../../../tslib/tslib');
var Lint = require('../index');
var Rule = (function(_super) {
  tslib_1.__extends(Rule, _super);
  function Rule() {
    return _super !== null && _super.apply(this, arguments) || this;
  }
  Rule.prototype.apply = function(sourceFile) {
    return this.applyWithFunction(sourceFile, walk);
  };
  Rule.metadata = {
    ruleName: "no-reference",
    description: "Disallows `/// <reference path=>` imports (use ES6-style imports instead).",
    rationale: (_a = ["\n            Using `/// <reference path=>` comments to load other files is outdated.\n            Use ES6-style imports to reference other files."], _a.raw = ["\n            Using \\`/// <reference path=>\\` comments to load other files is outdated.\n            Use ES6-style imports to reference other files."], Lint.Utils.dedent(_a)),
    optionsDescription: "Not configurable.",
    options: null,
    optionExamples: [true],
    type: "typescript",
    typescriptOnly: false
  };
  Rule.FAILURE_STRING = "<reference> is not allowed, use imports";
  return Rule;
}(Lint.Rules.AbstractRule));
exports.Rule = Rule;
function walk(ctx) {
  for (var _i = 0,
      _a = ctx.sourceFile.referencedFiles; _i < _a.length; _i++) {
    var ref = _a[_i];
    ctx.addFailure(ref.pos, ref.end, Rule.FAILURE_STRING);
  }
}
var _a;
