// tslint/lib/rules/noRequireImportsRule.js
/* */ 
"use strict";
Object.defineProperty(exports, "__esModule", {value: true});
var tslib_1 = require('../../../tslib/tslib');
var tsutils_1 = require('../../../tsutils/index');
var ts = require('../../../typescript/lib/typescript');
var Lint = require('../index');
var Rule = (function(_super) {
  tslib_1.__extends(Rule, _super);
  function Rule() {
    return _super !== null && _super.apply(this, arguments) || this;
  }
  Rule.prototype.apply = function(sourceFile) {
    return this.applyWithFunction(sourceFile, walk);
  };
  Rule.metadata = {
    ruleName: "no-require-imports",
    description: "Disallows invocation of `require()`.",
    rationale: "Prefer the newer ES6-style imports over `require()`.",
    optionsDescription: "Not configurable.",
    options: null,
    optionExamples: [true],
    type: "maintainability",
    typescriptOnly: false
  };
  Rule.FAILURE_STRING = "require() style import is forbidden";
  return Rule;
}(Lint.Rules.AbstractRule));
exports.Rule = Rule;
function walk(ctx) {
  return ts.forEachChild(ctx.sourceFile, function cb(node) {
    if (tsutils_1.isCallExpression(node)) {
      if (node.arguments.length !== 0 && tsutils_1.isIdentifier(node.expression) && node.expression.text === "require") {
        ctx.addFailureAtNode(node, Rule.FAILURE_STRING);
      }
    } else if (tsutils_1.isImportEqualsDeclaration(node) && node.moduleReference.kind === ts.SyntaxKind.ExternalModuleReference) {
      ctx.addFailureAtNode(node.moduleReference, Rule.FAILURE_STRING);
    }
    return ts.forEachChild(node, cb);
  });
}
