// jest-cli/build/generateEmptyCoverage.js
/* */ 
'use strict';
const IstanbulInstrument = require('../../istanbul-lib-instrument/dist/index');
var _require = require('../../jest-runtime/build/index');
const ScriptTransformer = _require.ScriptTransformer,
    shouldInstrument = _require.shouldInstrument;
module.exports = function(source, filename, globalConfig, config) {
  const coverageOptions = {
    collectCoverage: globalConfig.collectCoverage,
    collectCoverageFrom: globalConfig.collectCoverageFrom,
    collectCoverageOnlyFrom: globalConfig.collectCoverageOnlyFrom,
    mapCoverage: globalConfig.mapCoverage
  };
  if (shouldInstrument(filename, coverageOptions, config)) {
    const transformResult = new ScriptTransformer(config).transformSource(filename, source, false, globalConfig.mapCoverage);
    const instrumenter = IstanbulInstrument.createInstrumenter();
    instrumenter.instrumentSync(transformResult.code, filename);
    return {
      coverage: instrumenter.fileCoverage,
      sourceMapPath: transformResult.sourceMapPath
    };
  } else {
    return null;
  }
};
