// jest-config/build/defaults.js
/* */ 
(function(process) {
  'use strict';
  const os = require('os');
  const path = require('path');
  var _require = require('../../jest-regex-util/build/index');
  const replacePathSepForRegex = _require.replacePathSepForRegex;
  const constants = require('./constants');
  const NODE_MODULES_REGEXP = replacePathSepForRegex(constants.NODE_MODULES);
  const cacheDirectory = (() => {
    var _process = process;
    const getuid = _process.getuid;
    if (getuid == null) {
      return path.join(os.tmpdir(), 'jest');
    }
    return path.join(os.tmpdir(), 'jest_' + getuid.call(process).toString(36));
  })();
  module.exports = {
    automock: false,
    bail: false,
    browser: false,
    cache: true,
    cacheDirectory,
    clearMocks: false,
    coveragePathIgnorePatterns: [NODE_MODULES_REGEXP],
    coverageReporters: ['json', 'text', 'lcov', 'clover'],
    expand: false,
    globals: {},
    haste: {providesModuleNodeModules: []},
    mapCoverage: false,
    moduleDirectories: ['node_modules'],
    moduleFileExtensions: ['js', 'json', 'jsx', 'node'],
    moduleNameMapper: {},
    modulePathIgnorePatterns: [],
    noStackTrace: false,
    notify: false,
    preset: null,
    resetMocks: false,
    resetModules: false,
    snapshotSerializers: [],
    testEnvironment: 'jest-environment-jsdom',
    testMatch: ['**/__tests__/**/*.js?(x)', '**/?(*.)(spec|test).js?(x)'],
    testPathIgnorePatterns: [NODE_MODULES_REGEXP],
    testRegex: '',
    testResultsProcessor: null,
    testURL: 'about:blank',
    timers: 'real',
    transformIgnorePatterns: [NODE_MODULES_REGEXP],
    useStderr: false,
    verbose: null,
    watch: false,
    watchman: true
  };
})(require('process'));
