// tslint/lib/formatters/fileslistFormatter.js
/* */ 
"use strict";
Object.defineProperty(exports, "__esModule", {value: true});
var tslib_1 = require('../../../tslib/tslib');
var abstractFormatter_1 = require('../language/formatter/abstractFormatter');
var Formatter = (function(_super) {
  tslib_1.__extends(Formatter, _super);
  function Formatter() {
    return _super !== null && _super.apply(this, arguments) || this;
  }
  Formatter.prototype.format = function(failures) {
    if (failures.length === 0) {
      return "";
    }
    var files = [];
    var currentFile;
    for (var _i = 0,
        failures_1 = failures; _i < failures_1.length; _i++) {
      var failure = failures_1[_i];
      var fileName = failure.getFileName();
      if (fileName !== currentFile) {
        files.push(fileName);
        currentFile = fileName;
      }
    }
    return files.join("\n") + "\n";
  };
  Formatter.metadata = {
    formatterName: "filesList",
    description: "Lists files containing lint errors.",
    sample: "directory/myFile.ts",
    consumer: "machine"
  };
  return Formatter;
}(abstractFormatter_1.AbstractFormatter));
exports.Formatter = Formatter;
