// tslint/lib/language/rule/abstractRule.js
/* */ 
"use strict";
Object.defineProperty(exports, "__esModule", {value: true});
var walker_1 = require('../walker/index');
var AbstractRule = (function() {
  function AbstractRule(options) {
    this.options = options;
    this.ruleName = options.ruleName;
    this.ruleArguments = options.ruleArguments;
    this.ruleSeverity = options.ruleSeverity;
  }
  AbstractRule.prototype.getOptions = function() {
    return this.options;
  };
  AbstractRule.prototype.applyWithWalker = function(walker) {
    walker.walk(walker.getSourceFile());
    return walker.getFailures();
  };
  AbstractRule.prototype.isEnabled = function() {
    return this.ruleSeverity !== "off";
  };
  AbstractRule.prototype.applyWithFunction = function(sourceFile, walkFn, options) {
    var ctx = new walker_1.WalkContext(sourceFile, this.ruleName, options);
    walkFn(ctx);
    return ctx.failures;
  };
  AbstractRule.prototype.filterFailures = function(failures) {
    return failures;
  };
  return AbstractRule;
}());
exports.AbstractRule = AbstractRule;
