// istanbul-lib-hook/lib/hook.js
/* */ 
(function(process) {
  var path = require('path'),
      vm = require('vm'),
      appendTransform = require('../../append-transform/index'),
      originalCreateScript = vm.createScript,
      originalRunInThisContext = vm.runInThisContext;
  function transformFn(matcher, transformer, verbose) {
    return function(code, filename) {
      var shouldHook = typeof filename === 'string' && matcher(path.resolve(filename)),
          transformed,
          changed = false;
      if (shouldHook) {
        if (verbose) {
          console.error('Module load hook: transform [' + filename + ']');
        }
        try {
          transformed = transformer(code, filename);
          changed = true;
        } catch (ex) {
          console.error('Transformation error for', filename, '; return original code');
          console.error(ex.message || String(ex));
          if (verbose) {
            console.error(ex.stack);
          }
          transformed = code;
        }
      } else {
        transformed = code;
      }
      return {
        code: transformed,
        changed: changed
      };
    };
  }
  function unloadRequireCache(matcher) {
    if (matcher && typeof require !== 'undefined' && require && require.cache) {
      Object.keys(require.cache).forEach(function(filename) {
        if (matcher(filename)) {
          delete require.cache[filename];
        }
      });
    }
  }
  function hookRequire(matcher, transformer, options) {
    options = options || {};
    var extensions,
        disable = false,
        fn = transformFn(matcher, transformer, options.verbose),
        postLoadHook = options.postLoadHook && typeof options.postLoadHook === 'function' ? options.postLoadHook : null;
    extensions = options.extensions || ['.js'];
    extensions.forEach(function(ext) {
      appendTransform(function(code, filename) {
        if (disable) {
          return code;
        }
        var ret = fn(code, filename);
        if (postLoadHook) {
          postLoadHook(filename);
        }
        return ret.code;
      }, ext);
    });
    return function() {
      disable = true;
    };
  }
  function hookCreateScript(matcher, transformer, opts) {
    opts = opts || {};
    var fn = transformFn(matcher, transformer, opts.verbose);
    vm.createScript = function(code, file) {
      var ret = fn(code, file);
      return originalCreateScript(ret.code, file);
    };
  }
  function unhookCreateScript() {
    vm.createScript = originalCreateScript;
  }
  function hookRunInThisContext(matcher, transformer, opts) {
    opts = opts || {};
    var fn = transformFn(matcher, transformer, opts.verbose);
    vm.runInThisContext = function(code, file) {
      var ret = fn(code, file);
      return originalRunInThisContext(ret.code, file);
    };
  }
  function unhookRunInThisContext() {
    vm.runInThisContext = originalRunInThisContext;
  }
  module.exports = {
    hookRequire: hookRequire,
    hookCreateScript: hookCreateScript,
    unhookCreateScript: unhookCreateScript,
    hookRunInThisContext: hookRunInThisContext,
    unhookRunInThisContext: unhookRunInThisContext,
    unloadRequireCache: unloadRequireCache
  };
})(require('process'));
